src/software/ai/navigator/path_planner/theta_star_path_planner.h

```diff
@@ -40,25 +40,94 @@ class ThetaStarPathPlanner : public PathPlanner
                                  const std::vector<ObstaclePtr> &obstacles) override;
 
    private:
-    class Coordinate : public std::pair<unsigned int, unsigned int>
+    class Coordinate
     {
        public:
         Coordinate(unsigned int row, unsigned int col)
-            : std::pair<unsigned int, unsigned int>(row, col)
+            : row_(row), col_(col), internal_comparison_key_(computeComparisonKey(*this))
         {
         }
 
-        Coordinate() : std::pair<unsigned int, unsigned int>() {}
+        Coordinate() : row_(0), col_(0) {}
 
         unsigned int row(void) const
         {
-            return this->first;
+            return row_;
         }
 
         unsigned int col(void) const
         {
-            return this->second;
+            return col_;
         }
+
+        bool operator<(const Coordinate &other) const
+        {
+            return internal_comparison_key_ < other.internal_comparison_key_;
+        }
+
+        bool operator==(const Coordinate &other) const
+        {
+            return internal_comparison_key_ == other.internal_comparison_key_;
+        }
+
+        static unsigned long computeComparisonKey(const Coordinate &coord)
+        {
+            return coord.row() + coord.col() * ((1 << 15) + 1);
+        }
+
+       private:
+        unsigned int row_;
+        unsigned int col_;
+        unsigned long internal_comparison_key_;
+    };
+
+    class CoordinatePair
+    {
+       public:
+        CoordinatePair(const Coordinate &coord1, const Coordinate &coord2)
+            : coord1_(coord1),
+              coord2_(coord2),
+              internal_comparison_key_(computeComparisonKey(coord1, coord2))
+        {
+        }
+
+        CoordinatePair() : coord1_(), coord2_() {}
+
+        const Coordinate &firstCoordinate(void) const
+        {
+            return coord1_;
+        }
+
+        const Coordinate &secondCoordinate(void) const
+        {
+            return coord2_;
+        }
+
+        bool operator<(const CoordinatePair &other) const
+        {
+            return internal_comparison_key_ < other.internal_comparison_key_;
+        }
+
+        static unsigned long computeComparisonKey(const Coordinate &coord1,
+                                                  const Coordinate &coord2)
```

Comment:
javadocs and a comment explaining what's happening

Reply:
Added